did_peer/src/numalgos/numalgo2/resolve/helpers.rs

```diff
@@ -0,0 +1,413 @@
+use base64::{engine::general_purpose::STANDARD_NO_PAD, Engine};
+use did_doc::schema::{did_doc::DidDocumentBuilder, service::Service, types::uri::Uri, utils::OneOrList};
+use did_doc_sov::extra_fields::{aip1::ExtraFieldsAIP1, didcommv2::ExtraFieldsDidCommV2, ExtraFieldsSov};
+use did_parser::Did;
+
+use crate::{
+    error::DidPeerError,
+    key::{get_verification_methods_by_key, Key},
+    numalgos::numalgo2::{purpose::ElementPurpose, service_abbreviated::ServiceAbbreviated},
+    peer_did_resolver::options::PublicKeyEncoding,
+};
+
+pub fn process_elements(
+    mut did_doc_builder: DidDocumentBuilder<ExtraFieldsSov>,
+    did: &Did,
+    public_key_encoding: &PublicKeyEncoding,
+) -> Result<DidDocumentBuilder<ExtraFieldsSov>, DidPeerError> {
+    let numalgoless_id = did.id().chars().skip(1).collect::<String>();
+    let mut service_index: usize = 0;
+
+    // Skipping one here because the first element is empty string
+    for element in numalgoless_id.split('.').skip(1) {
+        did_doc_builder = process_element(element, did_doc_builder, &mut service_index, did, public_key_encoding)?;
+    }
+
+    Ok(did_doc_builder)
+}
+
+fn process_element(
+    element: &str,
+    mut did_doc_builder: DidDocumentBuilder<ExtraFieldsSov>,
+    service_index: &mut usize,
+    did: &Did,
+    public_key_encoding: &PublicKeyEncoding,
+) -> Result<DidDocumentBuilder<ExtraFieldsSov>, DidPeerError> {
+    let purpose: ElementPurpose = element
+        .chars()
+        .nth(0)
+        .ok_or(DidPeerError::DidValidationError(format!(
+            "No purpose code following element separator in '{}'",
+            element
+        )))?
+        .try_into()?;
+    let purposeless_element = element.chars().skip(1).collect::<String>();
+
+    if purpose == ElementPurpose::Service {
+        did_doc_builder = process_service_element(&purposeless_element, did_doc_builder, service_index)?;
+    } else {
+        did_doc_builder =
+            process_key_element(&purposeless_element, did_doc_builder, did, public_key_encoding, purpose)?;
+    }
+
+    Ok(did_doc_builder)
+}
+
+fn process_service_element(
```

Comment:
same note about naming, i'd suggest "mapping" or "encoding" vocabulary rather than "processing" across the board in this file

Reply:
Hmm, as I said, I prefer the current naming of "processing" of "elements", but of course my position is biased by coming up and being very used to this naming. Again, the notion "processing" of "elements" seems obvious to me given what is expected to resolve a `did:peer:2`.

Reply:
leaving final decision to your judgement